### PR TITLE
feat: adding failure retry

### PR DESCRIPTION
**Pull request by mayzhang2000**
Checklist:

* [ ] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [ ] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [ ] Optional. My organization is added to USERS.md.
* [ ] I've signed the CLA and my build is green ([troubleshooting builds](https://argoproj.github.io/argo-cd/developer-guide/ci/)). 

Ref: #2960

### CODE REVIEW

File: cmd/argocd-server/commands/root.go [hunk 1]
```diff
@@ -25,24 +27,26 @@ import (
...
+		staticAssetsDir           string
+		baseHRef                  string
+		rootPath                  string
+		repoServerAddress         string
+		dexServerAddress          string
+		disableAuth               bool
+		tlsConfigCustomizerSrc    func() (tls.ConfigCustomizer, error)
+		cacheSrc                  func() (*servercache.Cache, error)
+		frameOptions              string
+		failureRetryCount         int
```
Comment (alexmt) [inline cmd/argocd-server/commands/root.go:48]: Settings, that we don't expect the user to change, typically configured using environment variables. Can you please create two environment variables instead ? `ARGOCD_K8S_RETRY_COUNT` and `ARGOCD_K8S_RETRY_DURATION_MILLISECONDS`

File: cmd/argocd-server/commands/root.go [hunk 2]
```diff
@@ -25,24 +27,26 @@ import (
...
+		failureRetryPeriodSeconds int
```
Comment (alexmt) [inline cmd/argocd-server/commands/root.go:49]: 1 second between retries might be too much. Can you please change the unit to milliseconds and change default to 100 milliseconds.

File: controller/metrics/failureretrywrapper.go [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+}
+
+func (frt *failureRetryRoundTripper) RoundTrip(r *http.Request) (*http.Response, error) {
+	resp, roundTimeErr := frt.roundTripper.RoundTrip(r)
+	counter := frt.failureRetryCount
+	for shouldRetry(counter, r, resp, roundTimeErr) {
+		log.Debug("failureRetryRoundTripper: ", r.URL.Path, " ", r.Method)
+		resp, roundTimeErr = frt.roundTripper.RoundTrip(r)
+		counter--
+		time.Sleep(time.Duration(frt.failureRetryPeriodSeconds) * time.Second)
```
Comment (alexmt) [inline controller/metrics/failureretrywrapper.go:63]: Roundtripper unnecessary sleeps if retry attempt is successful. Can you please fix it?

**Review by alexmt — CHANGES_REQUESTED** [review]

File: cmd/argocd-server/commands/root.go [hunk 4]
```diff
@@ -4,6 +4,10 @@ import (
 	"context"
 	"time"
 
+	"github.com/argoproj/argo-cd/util/env"
```
Comment (alexmt) [inline cmd/argocd-server/commands/root.go:7]: Nitpick: please keep imports grouped.

**Review by alexmt — APPROVED** [review]
LGTM. 

Please group imports in failureretrywrapper.go and  root.go files. Typically we have three groups: imports from golang standard libraries, third-party imports, imports from argoproj.io/argo-cd

Example is in https://github.com/argoproj/argo-cd/blob/master/reposerver/server.go

**Comment by mayzhang2000** [comment]
fixes #2960

**Pull request merged**